### PR TITLE
Issue #8048: Extract logic for token type checking to separate method

### LINKED ISSUE #8048: Extract logic for token type checking to separate method
There is a recurring pattern throughout Checkstyle's code:
https://github.com/checkstyle/checkstyle/blob/a830b92f5ed1f9d2ed3b88440ba4d494105de89d/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java#L232-L235
https://github.com/checkstyle/checkstyle/blob/a830b92f5ed1f9d2ed3b88440ba4d494105de89d/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java#L141-L145
https://github.com/checkstyle/checkstyle/blob/a830b92f5ed1f9d2ed3b88440ba4d494105de89d/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java#L182-L185
https://github.com/checkstyle/checkstyle/blob/a830b92f5ed1f9d2ed3b88440ba4d494105de89d/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java#L208-L212
https://github.com/checkstyle/checkstyle/blob/8bf05ced07bb04c7240ae9f60ecda3c628e9b6f6/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/CommentsIndentationCheck.java#L409-L417
https://github.com/checkstyle/checkstyle/blob/8bf05ced07bb04c7240ae9f60ecda3c628e9b6f6/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/CommentsIndentationCheck.java#L424-L430
https://github.com/checkstyle/checkstyle/blob/bd45909e6345b2d3ef5c5472d35a3ad926a31eb9/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java#L738-L741

The main idea is that the AST node is being compared against a list of allowed types. What we can do is abstract away all of the `||` and `==` logic, into a new method that handles the comparisons for us. I see two ways of achieving this:

1. A static method under `TokenUtil`, `TokenUtil.isOfType(DetailAST ast, TokenTypes... types)`
https://github.com/checkstyle/checkstyle/blob/a830b92f5ed1f9d2ed3b88440ba4d494105de89d/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java#L232-L235
will become:
```java
if (TokenUtil.isOfType(parent, TokenTypes.CLASS_DEF, TokenTypes.INTERFACE_DEF, TokenTypes.ANNOTATION_DEF, TokenTypes.ENUM_DEF)) {
```
2. An instance method under `DetailAST`/`DetailASTImpl`, `ast.isOfType(TokenTypes... types)`
```java
if (parent.isOfType(TokenTypes.CLASS_DEF, TokenTypes.INTERFACE_DEF, TokenTypes.ANNOTATION_DEF, TokenTypes.ENUM_DEF)) {
```

### PR DESCRIPTION
**Pull request by shashwatj07**
Resolves #8048

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java [hunk 1]
```diff
@@ -269,4 +269,23 @@ public static boolean areOnSameLine(DetailAST ast1, DetailAST ast2) {
...
+     * Determines if the AST belongs to the given types.
+     *
+     * @param type the Token Type to check
+     * @param types the acceptable types
+     *
+     * @return true if type matches one of the given types.
+     */
+    public static boolean isOfType(int type, int... types) {
+        boolean result = false;
+        for (int t : types) {
```
Comment (wltan) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:282]: From https://github.com/checkstyle/checkstyle/issues/8048#issuecomment-618410435: I still think that a declarative programming approach with streams is a much more elegant implementation than this.

Comment (shashwat70) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:282]: I think you're right. That's a better implementation.

@romani @rnveach ping

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/TokenUtilTest.java [hunk 2]
```diff
@@ -277,4 +278,18 @@ public void testForEachChild() {
         assertEquals(secondSibling, firstChild, "Mismatched child node");
     }
 
+    @Test
+    public void testIsOfTypeTrue() {
```
Comment (timurt) [inline src/test/java/com/puppycrawl/tools/checkstyle/utils/TokenUtilTest.java:282]: it is better to create two separate tests with corresponding names

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java (hunk 1 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:282]: totally agree with @wltan

Comment (shashwat70) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java]: Done!

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/TokenUtilTest.java (hunk 2 above)
Comment (shashwat70) [reply src/test/java/com/puppycrawl/tools/checkstyle/utils/TokenUtilTest.java:282]: Done!

**Comment by shashwat70** [comment]
Taken https://github.com/checkstyle/checkstyle/issues/8048#issuecomment-619387931 into account!

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java [hunk 3]
```diff
@@ -269,4 +269,29 @@ public static boolean areOnSameLine(DetailAST ast1, DetailAST ast2) {
...
+     * Determines if the AST belongs to the given types.
+     *
+     * @param ast the AST node to check
+     * @param types the acceptable types
+     *
+     * @return true if type matches one of the given types.
+     */
+    public static boolean isOfType(DetailAST ast, int... types) {
+        return ast != null
+                && Arrays.stream(types).anyMatch(tokenType -> tokenType == ast.getType());
```
Comment (timurt) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:294]: please reuse function you wrote above

Comment (shashwat70) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java]: Done! @timurt

**Comment by timurt** [comment]
@romani please review

**Comment by romani** [comment]
@shashwat70 , please finish review.

**Comment by shashwat70** [comment]
> please refactor some cases that shown already in issue to use this method.
> PR, just to introduce methods is not that good, refactoring to use your methods is easy - please do.

I think it's better to refactor relevant lines file by file. I've it for one, i.e., ScopeUtil.java @romani

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java [hunk 4]
```diff
@@ -253,20 +243,18 @@ public static boolean isOuterMostType(DetailAST node) {
...
         // catch parameter?
-        if (node.getType() == TokenTypes.PARAMETER_DEF) {
+        if (TokenUtil.isOfType(node, TokenTypes.PARAMETER_DEF)) {
             final DetailAST parent = node.getParent();
-            localVariableDef = parent.getType() == TokenTypes.LITERAL_CATCH;
+            localVariableDef = TokenUtil.isOfType(parent.getType(), TokenTypes.LITERAL_CATCH);
         }
 
-        if (node.getType() == TokenTypes.RESOURCE) {
+        if (TokenUtil.isOfType(node, TokenTypes.RESOURCE)) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java:257]: please restore single node verification to use `==`, it is more concise and shorter and performance is better.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java [hunk 5]
```diff
@@ -70,18 +70,17 @@ public static Scope getSurroundingScope(DetailAST node) {
              token != null;
              token = token.getParent()) {
             final int type = token.getType();
-            if (type == TokenTypes.CLASS_DEF
-                || type == TokenTypes.INTERFACE_DEF
-                || type == TokenTypes.ANNOTATION_DEF
-                || type == TokenTypes.ENUM_DEF) {
+            if (TokenUtil.isOfType(type, TokenTypes.CLASS_DEF,
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java:73]: I'd like to have a dedicated method for Class/Interface/Enum/Annotation checking.
Something like `TokenUtil.isTypeDeclaration(tokenType)`

**Comment by romani** [comment]
@shashwatj07 , please finish PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java (hunk 4 above)
Comment (shashwatj07) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java (hunk 5 above)
Comment (shashwatj07) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java:73]: can we do it in a separate PR?

**Comment by shashwatj07** [comment]
> @shashwatj07 , please finish PR.

@romani done

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java (hunk 5 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java:73]: Sure.

Comment (shashwatj07) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java:73]: please open a new issue for it. i'll work on it as soon as it get approved.

**Comment by shashwatj07** [comment]
This shouldn't have failed, isn't it? @romani

**Comment by romani** [comment]
@shashwatj07 , there is conflict, please rebase.
I restarted Travis.

**Comment by shashwatj07** [comment]
> @shashwatj07 , there is conflict, please rebase.
> I restarted Travis.

Done. @romani

**Pull request merged**